service/internal/checksum/middleware_checksum_metrics_tracking.go

```diff
@@ -0,0 +1,90 @@
+package checksum
+
+import (
+	"context"
+	"fmt"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	awsmiddleware "github.com/aws/aws-sdk-go-v2/aws/middleware"
+	"github.com/aws/smithy-go/middleware"
+	smithyhttp "github.com/aws/smithy-go/transport/http"
+)
+
+var supportedChecksumFeatures = map[Algorithm]awsmiddleware.UserAgentFeature{
+	AlgorithmCRC32:     awsmiddleware.UserAgentFeatureRequestChecksumCRC32,
+	AlgorithmCRC32C:    awsmiddleware.UserAgentFeatureRequestChecksumCRC32C,
+	AlgorithmSHA1:      awsmiddleware.UserAgentFeatureRequestChecksumSHA1,
+	AlgorithmSHA256:    awsmiddleware.UserAgentFeatureRequestChecksumSHA256,
+	AlgorithmCRC64NVME: awsmiddleware.UserAgentFeatureRequestChecksumCRC64,
+}
+
+// RequestChecksumMetricsTracking is the middleware to track operation request's checksum usage
+type RequestChecksumMetricsTracking struct {
+	RequestChecksumCalculation aws.RequestChecksumCalculation
+	UserAgent                  *awsmiddleware.RequestUserAgent
+}
+
+// ID provides the middleware identifier
+func (m *RequestChecksumMetricsTracking) ID() string {
+	return "AWSChecksum:RequestMetricsTracking"
+}
+
+// HandleBuild checks request checksum config and checksum value sent
+// and sends corresponding feature id to user agent
+func (m *RequestChecksumMetricsTracking) HandleBuild(
+	ctx context.Context, in middleware.BuildInput, next middleware.BuildHandler,
+) (
+	out middleware.BuildOutput, metadata middleware.Metadata, err error,
+) {
+	req, ok := in.Request.(*smithyhttp.Request)
+	if !ok {
+		return out, metadata, fmt.Errorf("unknown request type %T", req)
+	}
+
+	switch m.RequestChecksumCalculation {
+	case aws.RequestChecksumCalculationWhenSupported:
+		m.UserAgent.AddUserAgentFeature(awsmiddleware.UserAgentFeatureRequestChecksumWhenSupported)
+	case aws.RequestChecksumCalculationWhenRequired:
+		m.UserAgent.AddUserAgentFeature(awsmiddleware.UserAgentFeatureRequestChecksumWhenRequired)
+	}
+
+	for algo, feat := range supportedChecksumFeatures {
+		checksumHeader := AlgorithmHTTPHeader(algo)
+		if checksum := req.Header.Get(checksumHeader); checksum != "" {
```

Comment:
I suspect you already checked this, but I'm surprised these headers are already set so early on the middleware stack

Reply:
they are set at [serialize](https://github.com/aws/aws-sdk-go-v2/blob/885de40869f9bcee29ad11d60967aa0f1b571d46/service/s3/serializers.go#L7729) step on stack

Reply:
the check here is only for checksums marked with feature ID